### PR TITLE
Revamp profiling and support for MPI timing

### PR DESCRIPTION
**Pull request by FabioLuporini**
This one can clearly wait until after holidays, but basically it enables MPI-level profiling. Rank-0 derives the global execution time as described [here](https://stackoverflow.com/a/5301414)

Potentially, this can be wired with `benchmark.py`

I'd like reviewers to actually **try** the code and take a look at the profiling output

### CODE REVIEW

**Comment by btrb** [comment]
hi! i'm not sure if that's what you meant but i hope it can be of use. i run:

`DEVITO_MPI=1 mpirun -n 4 python benchmark.py run -P acoustic -bm O2 -d 512 512 512 -so 12 -a basic --tn 100 2> log_1.txt`

the output is:

```
Allocating memory for vp(556, 290, 290)
Allocating memory for damp(534, 268, 268)
Allocating memory for src_coords(0, 3)
Allocating memory for src(34, 0)
Allocating memory for rec_coords(128, 3)
Allocating memory for src_coords(0, 3)
DSE: extract_time_invariants [flops: 93, elapsed: 0.01 s] >>
     eliminate_inter_stencil_redundancies [flops: 93, elapsed: 0.00 s] >>
     eliminate_intra_stencil_redundancies [flops: 93, elapsed: 0.00 s] >>
     factorize [flops: 63, elapsed: 0.04 s] >>
     finalize [flops: 63, elapsed: 0.00 s]
DSE: [Total elapsed: 0.05 s]
DLE: avoid_denormals (elapsed: 0.00 s)
     optimize_halospots (elapsed: 0.05 s)
     dist_parallelize (elapsed: 0.05 s)
     loop_blocking (elapsed: 0.10 s)
     simdize (elapsed: 0.02 s)
     minimize_remainders (elapsed: 0.02 s)
     hoist_prodders (elapsed: 0.02 s)
     [Total elapsed: 0.26 s]
Allocating memory for rec(34, 128)
Allocating memory for u(3, 556, 290, 290)
CustomCompiler: cache hit `/tmp/devito-jitcache-uid1000/6de48551a2505d747f689173cc7bee884a5c5a3c.c` [0.03 s]
AutoTuner: run <x0_blk0_size=8,y0_blk0_size=8,z0_blk0_size=8> took 4.541248 (s) in 5 timesteps
AutoTuner: run <x0_blk0_size=16,y0_blk0_size=16,z0_blk0_size=16> took 5.119356 (s) in 5 timesteps
AutoTuner: run <x0_blk0_size=24,y0_blk0_size=24,z0_blk0_size=24> took 5.318523 (s) in 5 timesteps
AutoTuner: run <x0_blk0_size=32,y0_blk0_size=32,z0_blk0_size=32> took 5.944450 (s) in 5 timesteps
AutoTuner: run <x0_blk0_size=64,y0_blk0_size=64,z0_blk0_size=64> took 5.990977 (s) in 5 timesteps
AutoTuner: run <x0_blk0_size=96,y0_blk0_size=96,z0_blk0_size=96> took 6.111177 (s) in 5 timesteps
AutoTuner: run <x0_blk0_size=128,y0_blk0_size=128,z0_blk0_size=128> took 6.492741 (s) in 5 timesteps
AutoTuner: run <x0_blk0_size=532,y0_blk0_size=266,z0_blk0_size=266> took 3.754168 (s) in 5 timesteps
AutoTuner: selected <x0_blk0_size=532,y0_blk0_size=266,z0_blk0_size=266>
Operator `Forward` run in 26.54 s
* section0<32,532,266,266> with OI=3.60 computed in 26.53 s [2.86 GFlops/s, 0.05 GPts/s]
* section1<<32,1>,<32,1>,<32,1>,<32,1>,<32,1>,<32,1>,<32,1>,<32,1>,<32,1>> with OI=8.69 computed in 0.00 s [0.06 GFlops/s, 0.00 GPts/s]
* section2<<32,512>,<32,512>,<32,512>,<32,512>,<32,512>,<32,512>,<32,512>,<32,512>,<32,512>,<32,512>> with OI=8.92 computed in 0.01 s [0.29 GFlops/s]
`Forward` configuration:  {'dse': 'advanced', 'dle': 'advanced'} 
```

**Comment by georgebisbas** [comment]
Hi @btrb , there is high chance that you got out of memory because currently mpi allocates `mpithreads \times grid_space` and this problem is large. You should also have in mind to avoid -bm 03 with gcc and tti if openmp is activated.
Could you please tell me if you have issues running this: `DEVITO_OPENMP=0  DEVITO_MPI=1 mpirun -n 4 python benchmark.py bench -P tti -bm O3 -d 112 102 90 -so 8 -a basic --tn 30`
and this:
`DEVITO_OPENMP=1  OMP_NUM_THREADS=2 DEVITO_MPI=1 mpirun -n 2 python benchmark.py bench -P tti -bm O2 -d 112 102 90 -so 8 -a basic --tn 30
`

**Comment by georgebisbas** [comment]
@FabioLuporini does that mean that always `timing_rank(0)>timing_rank(!0)` ?

**Comment by FabioLuporini** [comment]
@btrb do you wanna give this another spin?

**Comment by FabioLuporini** [comment]
it's now significantly different than it was when the PR was filed the first time. See #development on Slack

**Comment by mloubout** [comment]
While I understand and have no issue with the profiling beeing single rank (even if it is actually quite useful from experience on azure to have each rank print out AT/perf info and I would like to have it still as an options) I do think the actual debug part such as allocation should not be muted for any rank.

**Comment by FabioLuporini** [comment]
> even if it is actually quite useful from experience on azure to have each rank print out AT/perf info and I would like to have it still as an options

you still get the full log from each rank unless you `set_log_level(..., comm=my_mpi_comm)` (see for an example benchmark.py). The final information will however be redundant. 

>  I do think the actual debug part such as allocation should not be muted for any rank.
this hasn't changed, though I think it should be **enriched** like for example below

```
Allocating memory for rxx(2, 292, 552, 552) [global shape: (2, 552, 552, 552)]
```

As I said, that's for another day and therefore open to debate

**Comment by mloubout** [comment]
> The "final information" (i.e., what the Operator emits upon returning from apply) will however be redundant (i.e., the same stuff will be printed by each rank -- that's why I always run MPI with set_log_level(..., comm=my_mpi_comm))
I'm not sure I follow/get what you mean. Currently each rank prints its own performance, which is not the same for all, including diifferent AT shapes and different runtimes. Are you saying now it wil lall print the same thing?

**Comment by FabioLuporini** [comment]
> Are you saying now it wil lall print the same thing?

no, it will be like master, but:

* in addition, there will be **global** performance metrics
* all ranks will print, for example
```
section0[rank0] ...
...
section0[rank0] ...
```

but you made me realize I should improve this. Only rank0 should emit this sort of information, otherwise it's duplicated

**Comment by FabioLuporini** [comment]
> If possible I would like to see somewhere the problem name that is executed. In current setup, the submitted jobs will not have a clue in .e* or .o* what is the executed problem.

I will do that, but it's not the point of this PR

**Comment by FabioLuporini** [comment]
Merged.

**Comment by btrb** [comment]
> @btrb do you wanna give this another spin?

sorry just saw your msg!
ofc if that'd still be any helpful, plz let me know

File: devito/operator.py [hunk 1]
```diff
@@ -655,6 +655,11 @@ def __init__(self, grid, *args, **kwargs):
         super(ArgumentsMap, self).__init__(*args, **kwargs)
         self.grid = grid
 
+    @property
+    def comm(self):
+        """The MPI communicator the arguments are collective over."""
```
Comment (btrb) [inline devito/operator.py]: collected?

**Pull request merged**